cairo-contracts/core/src/channel/components/handler.cairo

```diff
@@ -0,0 +1,176 @@
+            let chan_end_on_b = self
+                .read_channel_end(@msg.packet.port_id_on_b, @msg.packet.chan_id_on_b);
+
+            self.recv_packet_validate(msg.clone(), chan_end_on_b.clone());
+
+            self.recv_packet_execute(msg, chan_end_on_b);
+        }
+    }
+
+    #[generate_trait]
+    pub(crate) impl RecvPacketImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        +Drop<TContractState>,
+        impl EventEmitter: ChannelEventEmitterComponent::HasComponent<TContractState>,
+        impl ClientHandler: ClientHandlerComponent::HasComponent<TContractState>,
+        impl RouterHandler: RouterHandlerComponent::HasComponent<TContractState>
+    > of RecvPacketTrait<TContractState> {
+        fn recv_packet_validate(
+            self: @ComponentState<TContractState>, msg: MsgRecvPacket, chan_end_on_b: ChannelEnd
+        ) {
+            msg.validate_basic();
+
+            assert(chan_end_on_b.is_open(), ChannelErrors::INVALID_CHANNEL_STATE);
+
+            assert(
+                chan_end_on_b
+                    .counterparty_matches(@msg.packet.port_id_on_a, @msg.packet.chan_id_on_a),
+                ChannelErrors::INVALID_COUNTERPARTY
+            );
+
+            // TODO: verify connection end if we ever decide to implement ICS-03
+
+            let host_height = get_block_number();
+
+            let host_timestamp = get_block_timestamp();
+
+            msg.packet.check_timed_out(@host_height, @host_timestamp);
+
+            let client_comp = get_dep_component!(self, ClientHandler);
+
+            let client = client_comp.get_client(chan_end_on_b.client_id.client_type);
+
+            let client_status = client.status(chan_end_on_b.client_id.sequence);
+
+            assert(client_status.is_active(), ChannelErrors::INACTIVE_CLIENT);
+
+            let client_latest_height = client.latest_height(chan_end_on_b.client_id.sequence);
+
+            msg.verify_proof_height(@client_latest_height);
+
+            let packet_commitment_on_a = msg.packet.compute_packet_commitment();
+
+            client
+                .verify_membership(
+                    chan_end_on_b.client_id.sequence,
+                    'commitment path', // TODO: implement commitment path
+                    packet_commitment_on_a,
+                    msg.proof_commitment_on_a
+                );
+        }
+        fn recv_packet_execute(
+            ref self: ComponentState<TContractState>, msg: MsgRecvPacket, chan_end_on_b: ChannelEnd
+        ) {
+            // Check if another relayer already relayed the packet.
+            match chan_end_on_b.ordering {
+                ChannelOrdering::Unordered => {},
+                ChannelOrdering::Ordered => { return; }
+            };
+
+            let router_comp = get_dep_component!(@self, RouterHandler);
+
+            let maybe_app_address = router_comp
+                .get_app_address(msg.packet.port_id_on_b.compute_key());
```

Comment:
I think the problem here is that the binding between channel and port should be permanent. So if we know the channel ID, we'd also know which port ID it should be. Similarly, the channel ID determines which counterparty channel/port the packet should contain, which should be validated. All these bindings should be contained in `ChannelEnd` already, so we can have a helper function to validate that using `ChannelEnd`.

Reply:
>I think the problem here is that the binding between channel and port should be permanent. So if we know the channel ID, we'd also know which port ID it should be. Similarly, the channel ID determines which counterparty channel/port the packet should contain, which should be validated

On the same page with you. But these are already being validated IIUC. Once a receiving packet is opened, we read the respective `ChannelEnd` from the storage. Then, we either explicitly like [here](https://github.com/informalsystems/ibc-starknet/blob/farhad/impl-channel/cairo-contracts/core/src/channel/components/handler.cairo#L70-L74) or implicitly through [verify_membership](https://github.com/informalsystems/ibc-starknet/blob/farhad/impl-channel/cairo-contracts/core/src/channel/components/handler.cairo#L99) perform the checks. By the way, I'll go over the `recv_packet_validate` once more when everything is settled to optimize the implementation around.

> All these bindings should be contained in ChannelEnd already, so we can have a helper function to validate that using ChannelEnd.

I'm not sure I follow. Are you thinking of a different version of the `ChannelEnd` struct? Right now, I’m just keeping things simple and sticking close to how ibc-go and ibc-rs do it. But there's definitely room for improvement once things came together and stabilized a bit.

Reply:
It seems like the validation only checks whether the counterparty channel/port matches, but not whether the port ID matches the one that the channel is bound to.

> Are you thinking of a different version of the `ChannelEnd` struct?

Yeah, it doesn't really need to be the `ChannelEnd` struct itself. I am thinking more like a channel-specific context for handling a packet. For example, something like:

```rust
struct ChannelPacketHandler {
  pub local_channel_id: LocalChannelId,
  pub local_port_id: LocalPortId,
  pub counterparty_channel_id: CounterpartyChannelId,
  pub counterparty_port_id: CounterpartyPortId,
  pub receive_packet_commitment_storage: Map<Sequence, PacketCommitment>,
  // or
  pub storage: ContractState,
  // etc
}
```

The exact layout of the context doesn't really matter. It is just that if we can pack all the information we have available from the start, and then do the handling, it could make the code cleaner.

Reply:
Okay. Got it what you mean now.

> but not whether the port ID matches the one that the channel is bound to.

This also gets checked when retrieving the `ChannelEnd` from the store, since it's done using a pair of `(PortId, ChannelId)`.

I’ve cleaned up the code to make it more streamlined and closer to what you had in mind. In terms of structure, I actually tend to keep things consistent with the current IBC specs **wherever possible**, so this `ChannelEnd` struct is modeled after ICS-04 and the provable store format in ICS-24. 
Based on what you mentioned, I think I've hit the mark. If not, just let me know! I’ll dig into it in the next PR when I’ll be consolidating the implementation by adding tests.

Reply:
> This also gets checked when retrieving the ChannelEnd from the store

Ah alright then. Missed that part of the implicit logic.

> I actually tend to keep things consistent with the current IBC specs wherever possible

Ok, let's finish the implementation first, then we will look at if things can be improved.